### PR TITLE
Implement ObjectRewriter.

### PR DESCRIPTION
**Pull request by coryan**
This is part of the work for #816. It introduces a new class to manage
rewrite loops. The `Objects: rewrite` API may require multiple calls
before the (logical) operation completes. The ObjectRewriter offers a
simple member function to run the iteration until it completes. For
applications that want to be notified of progress the class offers a
member function with a callback parameter. Finally the application can
take full control and call `Iterate()` explicitly.

This PR contains integration and unit tests for this class, but we do
not expect application developers to construct it directly. A future PR
will introduce the `Client::RewriteObject` member function to return a
`ObjectRewriter`. In that PR we will introduce the examples and change
both unit and integration tests to use the higher-level API.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1154)

### CODE REVIEW

File: google/cloud/storage/object_rewriter.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+   * @note This operation blocks until the copy is finished. For very large
+   * objects that could take substantial time. Applications may need to persist
+   * the rewrite operation. Some applications may want to wrap this call with
+   * `std::async`, and run the copy on a separate thread.
+   *
+   * @return the object metadata once the copy completes.
+   */
+  ObjectMetadata Result() {
+    return ResultWithProgressCallback([](RewriteProgress const&) {});
+  }
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_rewriter.h]: > When surfaced in client.h, will the surface design expect developers to call Result() directly or will the RPC handle this for them? I'd find it easier to use rewrite RPC as a blocking call which waits until all bytes are copied. I'm also just making sure.

The prototype will be `ObjectRewriter Client::RewriteObject()`.  The developers can chain these things: `client.RewriteObject().Result()`, or we could add a prototype for them.  I modeled this after (some) of the other libraries.

> How would a developer call std::async?

Something like:

```C++
auto rewriter = client.RewriteObject(....);
auto future = std::async(std::launch::async, &ObjectRewriter::Result, rewriter);
// ... do some stuff in this thread, when ready to block call:
ObjectMetadata meta = future.get();
```

There are other ways to do it, like:

```C++
auto future = std::async(std::launch::async,
     [](gcs::Client client) { return client.RewriteObject(...).Result(); }, client);
```

File: google/cloud/storage/object_rewriter.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+   *
+   * @note This operation blocks until the copy is finished. For very large
+   * objects that could take substantial time. Applications may need to persist
+   * the rewrite operation. Some applications may want to wrap this call with
+   * `std::async`, and run the copy on a separate thread.
+   *
+   * @param cb the callback object.
+   *
+   * @tparam Functor the type of the callback object. It must satisfy:
+   *   `std:is_invocable<Functor, RewriteProgress>:: value == true`.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/object_rewriter.h]: Yeah, so Doxygen gets confused and thinks `value` is a link to a function if it shows up right after the colons:

https://www.stack.nl/~dimitri/doxygen/manual/autolink.html#linkfunc

File: google/cloud/storage/tests/object_integration_test.cc [hunk 3]
```diff
@@ -649,16 +649,42 @@ TEST_F(ObjectIntegrationTest, ComposedUsingEncryptedObject) {
...
+  auto bucket_name = ObjectTestEnvironment::bucket_name();
+  auto source_name = MakeRandomObjectName();
+
+  // Create the object, but only if it does not exist already.
+  ObjectMetadata source_meta = client.InsertObject(
+      bucket_name, source_name, LoremIpsum(), IfGenerationMatch(0));
+  EXPECT_EQ(source_name, source_meta.name());
+  EXPECT_EQ(bucket_name, source_meta.bucket());
+
+  // Compose new of object using previously created object
```
Comment (houglum) [inline google/cloud/storage/tests/object_integration_test.cc]: s/new of/new/

**Comment by frankyn** [comment]
Thanks, this helps! Clang has a very helpful message. It will depend on the compiler used, but overall it will be easier to understand in the future.

**Pull request merged**